utils/common/src/main/java/brooklyn/util/crypto/AuthorizedKeysParser.java

```diff
@@ -0,0 +1,132 @@
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package brooklyn.util.crypto;
+
+import java.io.ByteArrayInputStream;
+import java.io.ByteArrayOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.math.BigInteger;
+import java.security.KeyFactory;
+import java.security.PublicKey;
+import java.security.interfaces.DSAPublicKey;
+import java.security.interfaces.RSAPublicKey;
+import java.security.spec.DSAPublicKeySpec;
+import java.security.spec.RSAPublicKeySpec;
+
+import brooklyn.util.exceptions.Exceptions;
+
+import com.google.common.io.BaseEncoding;
+
+public class AuthorizedKeysParser {
+
+    public static PublicKey decodePublicKey(String keyLine) {
+        try {
+            ByteArrayInputStream stream = null;
+
+            // look for the Base64 encoded part of the line to decode
+            // both ssh-rsa and ssh-dss begin with "AAAA" due to the length bytes
+            for (String part : keyLine.split(" ")) {
+                if (part.startsWith("AAAA")) {
+                    stream = new ByteArrayInputStream(BaseEncoding.base64().decode(part));
+                    break;
+                }
+            }
+            if (stream == null)
+                throw new IllegalArgumentException("Encoded public key should include phrase beginning AAAA.");
+
+            String type = readType(stream);
+            if (type.equals("ssh-rsa")) {
+                BigInteger e = readBigInt(stream);
+                BigInteger m = readBigInt(stream);
+                RSAPublicKeySpec spec = new RSAPublicKeySpec(m, e);
+                return KeyFactory.getInstance("RSA").generatePublic(spec);
+            } else if (type.equals("ssh-dss")) {
+                BigInteger p = readBigInt(stream);
+                BigInteger q = readBigInt(stream);
+                BigInteger g = readBigInt(stream);
+                BigInteger y = readBigInt(stream);
+                DSAPublicKeySpec spec = new DSAPublicKeySpec(y, p, q, g);
+                return KeyFactory.getInstance("DSA").generatePublic(spec);
+            } else {
+                throw new IllegalArgumentException("Unknown public key type " + type);
+            }
+            
+        } catch (Exception e) {
+            Exceptions.propagateIfFatal(e);
+            throw new IllegalArgumentException("Error parsing authorized_keys/SSH2 format public key: "+e);
+        }
+    }
+
+    private static int readInt(InputStream stream) throws IOException {
+        return ((stream.read() & 0xFF) << 24) | ((stream.read() & 0xFF) << 16)
+            | ((stream.read() & 0xFF) << 8) | (stream.read() & 0xFF);
+    }
+    
+    private static byte[] readBytesWithLength(InputStream stream) throws IOException {
+        int len = readInt(stream);
+        byte[] result = new byte[len];
```

Comment:
Perhaps guard this - if `len` is big then there's something wrong (given the context of this method). Better to throw a graceful exception rather than cause an OutOfMemoryError.